### ISSUE TITLE
Missing preprocess.py file mentioned in README

### ISSUE THREAD

**Issue by wwwmj23** [dongdongtong/C3R_UDA#1]
Hi, thank you for sharing this great work! I have been trying to reproduce the results and encountered an issue related to the data preprocessing step.

In the README, it mentions running preprocess.py to handle the dataset. However, I could not find this file in the repository.

**Comment by dongdongtong** [comment]
Dear wwwmj23,

Thanks for your interest on our work, sorry for the late response.

For access convenience, we have uploaded the entire preprocessed data in this repo, please double check the Readme and download the preprocessed dataset.

We may upload the `preprocess.py` later.

If you have any related questions about the dataset, feel free to reopen this issue.
For other questions, please open a new issue.

BTW, if you want to reproduce the results, please create the python env following the `Readme`, we have tested in higher Pytorch versions but the results got worse.

Yours sincerely,
Shaodong

**Comment by wwwmj23** [comment]
> Dear wwwmj23,
> 
> Thanks for your interest on our work, sorry for the late response.
> 
> For access convenience, we have uploaded the entire preprocessed data in this repo, please double check the Readme and download the preprocessed dataset.
> 
> We may upload the `preprocess.py` later.
> 
> If you have any related questions about the dataset, feel free to reopen this issue. For other questions, please open a new issue.
> 
> BTW, if you want to reproduce the results, please create the python env following the `Readme`, we have tested in higher Pytorch versions but the results got worse.
> 
> Yours sincerely, Shaodong

Thank you very much for your prompt reply.

We have successfully used the raw abdomen dataset and the preprocessing code you uploaded. However, after running the load_data_wy.py, we noticed that the number of generated samples is different from the length of the data list in the train_ct.txt file mentioned for the SIFA method.

To ensure we are aligned with your experiments, would it be possible for you to upload the final preprocessed abdomen dataset? This would be a great help to us.

Thank you for your time and consideration.

**Comment by dongdongtong** [comment]
Dear wwwmj23,

The processing script of abdomen is written by ourself, so the final training length may not be consistent with SIFA.

However, we derive the final processed abdominal dataset using the same train split and test split, consistent with SIFA.

Our work is conducted based on our processed dataset using the `preprocess.py`. Please do not worry about this if you prepare to start a new work.

We believe that the most elegant way to read a dataset is to read it using its original data structure, not to preprocess it at first and then load it using another reading logic. If a new work starts, we just download the original dataset and read it, enough.

Yours sincerely,
Shaodong

**Comment by wwwmj23** [comment]
> Dear wwwmj23,
> 
> The processing script of abdomen is written by ourself, so the final training length may not be consistent with SIFA.
> 
> However, we derive the final processed abdominal dataset using the same train split and test split, consistent with SIFA.
> 
> Our work is conducted based on our processed dataset using the `preprocess.py`. Please do not worry about this if you prepare to start a new work.
> 
> We believe that the most elegant way to read a dataset is to read it using its original data structure, not to preprocess it at first and then load it using another reading logic. If a new work starts, we just download the original dataset and read it, enough.
> 
> Yours sincerely, Shaodong

Hi, thanks for your kind clarification.

For fairness and to make a consistent comparison with your paper, we still think it is better to follow the original dataset split in SIFA.

Specifically, we have prepared and processed the datasets as follows:

CT: training set 2400 slices, validation set 600 slices, test set composed of cases 1/9/30/32/33/39 (.npz format, similar to the heart dataset nii.gz.npz).

MR: training set 1600 slices, validation set 400 slices, test set composed of cases 1/13/32/38 (.npz format, similar to the heart dataset nii.gz.npz).

Both CT and MR training/validation are normalized. The min–max values we used are: MR: (-1.3, 4.3), CT: (-1.4, 3.6).

However, our reproduced results of MPSCL are:

CT → MR: Dice = 81.2

MR → CT: Dice = 76.5

while in your paper, the reported results are:

CT → MR: Dice = 86.61

MR → CT: Dice = 82.24.

Could you please kindly give us some suggestions on how to further align our experimental setting with yours? Do you think there might be something wrong in our preprocessing or normalization steps?

Thanks a lot for your help!

**Comment by wwwmj23** [comment]
In the line

for ii in range(int(np.floor(data.shape[2] // BATCHSIZE))):


may I ask what value of BATCHSIZE you used for the abdominal dataset?

**Comment by dongdongtong** [comment]
Ok.

For the misalignment between the reproduced performance of MPSCL, we also met some trouble in its reproducing. I suggest that you should freeze the batch norm layers, because BN tries to use the mean/std statistics of the whole dataset including both MR and CT, which potentially impacts the UDA training. Note that there exits intensity shift between MR and CT. We know that It is quite annoying.
Also, you may need to tune the hyperparameters, such as learning rate, weight decay, the trade-off between training losses, also the momentum updating speed of prototypes of MPSCL.

For the min-max of the abdominal, we directly use the min and max values calculated by the `preprocess.py` script. If you use them, fine.

For the Batch size, it is actually 8.

And if my remember is correct, SIFA does not have a test split in abdominal, please see [#issues51](https://github.com/cchen-cc/SIFA/issues/51) of SIFA.

**Comment by wwwmj23** [comment]
Hi, shaodong,

Thank you again for your help and guidance. Following your advice, we have adjusted the normalization values based on the numerical range of our training dataset.

We were hoping to ask for your insights on hyperparameter tuning as we continue to optimize the model. Specifically, could you share your experience on how you adjusted parameters such as the learning rate, weight decay, the trade-off between different training losses, and the momentum update speed for the MPSCL prototypes to achieve the best results?

Additionally, regarding the dataset setup, you asked in the SIFA issue:

"I found the val.txt contains the abdominal test cases (i.e., CT 1/9/30/32/33/39 and MRT2 1/13/32/38/). Does it mean that you evaluate your 2D adapted models using the average 2D Dice of the augmented axial slices in the val.txt other than the average 3D Dice of 3D nii.gz files like the cardiac evaluation?"

Could you please clarify how you ultimately configured the final test set for evaluation? Did you use the average 2D Dice of the axial slices, or was the evaluation based on the 3D Dice of the 3D nii.gz volumes?

Thank you again for your valuable time and assistance. We look forward to your reply.

**Comment by dongdongtong** [comment]
Hi, wwwmj23,

I really want to help detail the hyperparameter tuning, however, since there exits two-stage of MPSCL (and the second depends heavily on the first stage), our experience may not help you. But we can give some notes, for example, you should pay attention to the discriminator type between `PatchGAN` and `Orig_D` of MPSCL. Also, the lambda of adversarial loss, the lambda of contrastive learning is important and you should set as smaller as possible (i.e., 1e-5, 1e-4, 1e-3). Moreover, the training of the discriminator should be slower than generator, which means that you should always set a smaller learning rate for D than G.
Since I am not the expert of MPSCL, you may consult the author of `MPSCL`. They also maintain a repo on github.

For Q2, we do not know what kinds of evaluation adopted by SIFA, but ours is evaluated volume-wise, not averaged on slices.

**Comment by wwwmj23** [comment]
I would like to clarify one more point about the abdominal test set used in your evaluation. Is the final abdominal test volume you used identical to the officially downloaded data, or was it additionally processed before evaluation?

In the SIFA paper, the authors mention several preprocessing steps for the abdominal dataset, such as removing axial slices that do not contain target organs, cropping the black background, and resizing to 256×256. Could you please explain whether your test set was preprocessed in the same way (except for augmentation)?

If yes, would it be possible to share the corresponding preprocessing script/code so that we can reproduce the evaluation protocol? If the code is not convenient to release, could you briefly describe the exact preprocessing pipeline you applied to the test set? Alternatively, if sharing the code is not possible, would you be able to share the final processed abdominal test data used for evaluation (e.g., the processed volumes/files)? 

Thank you again for your time and assistance!

**Comment by dongdongtong** [comment]
Dear wwwmj23,

If my remember is right, we do not use other post-process teches to polish the test data. 

Right now, we have re-run the evaluate code and everything goes well, we have uploded the `evaluate.py` and you could fetch it here.

And we call the evaluate function like follows:
```
from datasets.load_data_wy import get_CT_data, get_MRI_data

# _,_,ct_val_X,ct_val_Y,ct_val_spacing, ct_min_inten, ct_max_inten = get_CT_data()
_,_,mr_val_X,mr_val_Y,mr_val_spacing, mr_min_inten, mr_max_inten = get_MRI_data()

# suppose we have trained the CT to MR model
model.load_state_dict(torch.load("/path/to/your/model.pt")))

model = model.cuda()
model.eval()
with torch.no_grad():
    # evaluate_ct(model, ct_val_X, ct_val_Y, use_assd=True)
    evaluate_mr(model, mr_val_X, mr_val_Y, use_assd=True, use_mask=False, use_hd=False, val_spacing=mr_val_spacing)

```

By the way, the script `load_data_wy.py` could be found in the downloaded data from our shared dataset link.

[evaluate.py](https://github.com/user-attachments/files/24593971/evaluate.py)
